src/shujinosuke.ts

```diff
@@ -0,0 +1,553 @@
+  } else if (isEvent(e)) {
+    const event = JSON.parse(e.postData.contents).event as SlackEvent;
+    handleSlackEvent(client, event);
+  }
+  return ContentService.createTextOutput('OK');
+}
+
+const handleSlackAction = (client, payload: SlackAction) => {
+  switch (payload.type) {
+    case 'block_actions':
+      handleBlockAction(client, payload)
+  }
+}
+
+const handleBlockAction = (client, payload: BlockAction) => {
+  const buttons = payload.actions.filter(action => action.type === 'button') as ButtonAction[];
+  if ('join' in buttons.map(action => action.value)) {
+    join(client, payload.channel.id, payload.user.id);
+  }
+}
+
+
+const getHelpMessage = (channelId: string) => {
+  const helpCommandsInWaiting = {
+    会議の開始: "`開始`",
+    Botステータスの確認: "`status`",
+    ping: "`ping`",
+    ヘルプ: "`ヘルプ` `help`",
+    アクティブメンバーの確認: "`誰いる？` `今いる人は？`",
+  };
+  const helpCommandsInMeeting = {
+    レポートの投稿: "`レポート` `調子、出来事、悩み等`",
+    会議の開始: "`開始`",
+    会議の強制終了: "`終了` `リセット` `reset`",
+    会議へ参加: "`参加`",
+    参加の取り消し: "`キャンセル`",
+    レポート未投稿者の確認: "`誰？` `残りは？`",
+    Botステータスの確認: "`status`",
+    ping: "`ping`",
+    ヘルプ: "`ヘルプ` `help`",
+    アクティブメンバーの確認: "`誰いる？` `今いる人は？`",
+  };
+
+  const channelState = getChannelState(channelId);
+  if (channelState) {
+    const commands = Object.entries(helpCommandsInWaiting)
+      .map(([command, explanation]) => command + "\n" + explanation)
+      .join("\n\n");
+    return (
+      `:books:会議開始前にShujinosukeで使えるコマンドは以下の通りです！\n` +
+      `:bulb:コマンドの前には必ず "@Shujinosuke" をつけましょう！\n\n` +
+      `${commands}`
+    );
+  } else {
+    const commands = Object.entries(helpCommandsInMeeting)
+      .map(([command, explanation]) => command + "\n" + explanation)
+      .join("\n\n");
+
+    return (
+      `:books:会議中にShujinosukeで使えるコマンドは以下の通りです！\n` +
+      `:bulb:コマンドの前には必ず "@Shujinosuke" をつけましょう！\n\n` +
+      commands
+    );
+  }
+}
+
+const getChannelStateMessage = (channelId: string) => {
+  const channelState = getChannelState(channelId);
+  if (channelState) {
+    return (
+      `\`\`\`\n` +
+      `${JSON.stringify(channelState, null, 2)}\n` +
+      `\`\`\``
+    );
+  } else {
+    return (
+      `\`\`\`\n` +
+      `{}\n` +
+      `\`\`\``
+    )
+  }
+}
+
+const getListen = (client: SlackClient, event: SlackEvent) => {
+  switch (event.type) {
+    case 'app_mention':
+      return (regExp: RegExp, callback: (client: SlackClient, event: AppMentionEvent) => void) => {
+        const messageContent = event.text.replace(/^<@\w+>\s*/, '');
+        if (messageContent.match(regExp)) {
+          callback(client, event);
+        }
+      }
+  }
+}
+const getThreadTs = (event: GenericMessageEvent | AppMentionEvent) => {
+  return event.thread_ts ? event.thread_ts : event.ts;
+}
+
+
+const handleSlackEvent = (client: SlackClient, event: SlackEvent) => {
+  switch (event.type) {
+    case 'app_mention':
+      handleAppMention(client, event as AppMentionEvent);
+      break;
```

Comment:
つまりここで`event.type`だけを見ると"app_mention"を新規と編集に分離できないということかな。

型情報じゃなくてAPI docの方で恐縮だけど、 https://api.slack.com/events/app_mention によると
> The app_mention event is not a message like the message.* event types. However, your app can treat its contents similarly.

とあり、 https://api.slack.com/events/message を見れば
> If the message has been edited after posting it will include an edited property, including the user ID of the editor, and the timestamp the edit happened. 

とあるので、eventの大本のペイロードには`.edited`が載っており、それで判断できるんじゃないかと思う。それとも大本のペイロードの時点でeditedは載ってない？

editedが載ってるんであれば、`AppMentionEvent`型に`as`でconformさせるとその情報が失われるというのがここでのblockerなんで、
https://github.com/slackapi/java-slack-sdk/blob/main/slack-api-model/src/main/java/com/slack/api/model/event/AppMentionEvent.java （コード生成元がjavaなのでそっちのリンク）
docの記述を信じてこのイベントをMessage型（GenericMessageEvent型）として扱えば大丈夫なんじゃなかろうか。
https://github.com/slackapi/java-slack-sdk/blob/main/slack-api-model/src/main/java/com/slack/api/model/Message.java

editedがeventにそもそも載ってないんであればしょうがないですが。

Reply:
TypeScript版 Slack Boltの型情報から、編集はmessage subtypeのchanged_eventからしか判定できないと思い込んでしまっていましたが、ログを確認したところ確かにeditedプロパティも含まれる様です。
（公式ドキュメントを見る癖はついてきていても「読む」癖がまだまだ足りてないようです...精進します...）


https://github.com/slackapi/bolt-js/blob/88e67902b3e7a6f60c8da1148f08ab698434b969/src/types/events/base-events.ts#L146 によると、AppMentionEventとMessageEventの違いは、eventのtypeが異なるのと、イベントの発生時刻が`event_ts` か`ts`で異なるという２つのようです。

---

118f1f5ca0be143ede6bbc6b0260d6ad4da942f5 でGenericMessageEvent型を利用するように変更してみました。
TypeScriptの型の扱いについて、まだ初歩的な知識しかないので愚直に置き換えています。
（propのevent, event_tsを利用していないので処理への影響は今のところない）

Reply:
ソースコメントも含めナイスです。
今週これはもうデプロイされてましたか？ 動作は問題ありませんでした！

Reply:
ありがとうございます！良かったです！

今週は１つ前のコミット( 487c11cf9274ce219e17ef1bf51c054ca3e25c49 ) の状態でデプロイされていました。
最新のコミットの動作確認は先ほど試し、レポートを編集してもメッセージが再送されないことは確認してあります。